plugin.video.carnet-meduza/resources/lib/carnet-meduza.py

```diff
@@ -0,0 +1,264 @@
+	videos = request_search_videos.json()
+	return videos 
+
+def category_video_count(category_id):
+	"""Returns total number of videos inside the category"""
+	request_category_video_count = requests.get(api_base_url + 'category/count/?id=' + category_id + '&uid=' + api_key)
+	category_video_count = request_category_video_count.json()
+	return category_video_count
+
+def video_url_description(video_id):
+	"""Returns video description"""
+	request_video_url_description = requests.get(api_base_url + 'video/?id=' + video_id + '&uid=' + api_key).json()
+	extract_keys = ['stream_url', 'opis']
+	video_url_description = dict((k, request_video_url_description[k]) for k in extract_keys if k in request_video_url_description)
+	return video_url_description
+
+def recommended_videos():
+	"""Returns list of recommended videos. The number of videos can be defined inside settings. Default is 20."""
+	num_recommends = xbmcaddon.Addon('plugin.video.carnet-meduza').getSetting('num_recommends')
+	request_recommended_videos = requests.get(api_base_url + 'recommended/?number=' + str(num_recommends) + '&uid=' + api_key)
+	videos = request_recommended_videos.json()
+	return videos 
+
+def list_search_or_recommended_videos(videos):
+	if videos == None:
+		return
+	else:	
+		for video in videos:
+			category_id = video['ID_kategorija']
+			name = video['naslov']
+			video_id = video['ID']
+			genre = video['kategorija']
+			duration = reduce(lambda x, y: x*60+y, [int(i) for i in (video['trajanje'].replace(':',',')).split(',')])
+			video_info = video_url_description(video_id)
+			try:
+				url = video_info['stream_url'].encode('utf-8') + '|User-Agent=Mozilla/5.0&Referer=https://meduza.carnet.hr'
+				description = video_info['opis'].encode('utf-8')
+			except KeyError:
+				url = ''
+				description = addon.getLocalizedString(30207)
+				pass		
+			image = video['slika']
+			li = xbmcgui.ListItem(name, iconImage=image)
+			li.setInfo( type="Video", infoLabels={ 
+							"Plot": description, 
+							"Genre": genre,
+							"Duration": duration
+								})
+			xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=li)
+		xbmcplugin.endOfDirectory(addon_handle)
+	
+def list_category_videos(videos,current_page_number,foldername):
+	if not videos:
+		li_translate = addon.getLocalizedString(30208)
+		li = xbmcgui.ListItem(li_translate, iconImage='DefaultVideo.png')
+		xbmcplugin.addDirectoryItem(handle=addon_handle, url='', listitem=li)
+		xbmcplugin.endOfDirectory(addon_handle)
+	else:
+		for video in videos:
+			category_id = video['ID_kategorija']
+			video_num = category_video_count(category_id)['count']
+			name = video['naslov']
+			video_id = video['ID']
+			duration = reduce(lambda x, y: x*60+y, [int(i) for i in (video['trajanje'].replace(':',',')).split(',')])
+			video_info = video_url_description(video_id)
+			url = video_info['stream_url'] + '|User-Agent=Mozilla/5.0&Referer=https://meduza.carnet.hr'
```

Comment:
I'm not sure of the usage here, but when adding header values, they should be urlencoded. (I don't know if that's happening elsewhere, but not doing the encoding can lead to some really tough issues to track down)